### DISCUSSION TITLE
What is the purpose of the `nu-complete` items in the config.nu file?

### DISCUSSION THREAD

**Discussion by wmstack** [nushell/nushell#5930] [Q&A] [answered]
I can't understand what they do. I thought they enable you to complete certain git instructions, however I had no success trying to do the `git branches` or `git remotes` instructions. Nothing magical happened, and git told me those commands didn't exist. How do you use those configuration options?

**Comment by wmstack** [comment]
I have figured it out. You need to export those items in order for them to be activated and become usable.

```nu
export def "nu-complete git branches" [] {
    ^git branch | lines | each { |line| $line | str replace '[\*\+] ' '' | str trim }
}

export def "nu-complete git remotes" [] {
    ^git remote | lines | each { |line| $line | str trim }
}
```

**Comment by fdncred** [comment, marked as answer]
Actually, no. You're not supposed to put `export` on those two custom commands. I mean that's true that you can do that to run them individually but they're not there to be used on their own. They're part of completions. Notice how they're used in the custom completion for `git fetch`.

```
module completions {
  # Custom completions for external commands (those outside of Nushell)
  # Each completions has two parts: the form of the external command, including its flags and parameters
  # and a helper command that knows how to complete values for those flags and parameters
  #
  # This is a simplified version of completions for git branches and git remotes
  def "nu-complete git branches" [] {
    ^git branch | lines | each { |line| $line | str replace '[\*\+] ' '' | str trim }
  }

  def "nu-complete git remotes" [] {
    ^git remote | lines | each { |line| $line | str trim }
  }

  # Download objects and refs from another repository
  export extern "git fetch" [
    repository?: string@"nu-complete git remotes" # name of the repository to fetch
    branch?: string@"nu-complete git branches" # name of the branch to fetch
    --all                                         # Fetch all remotes
    --append(-a)                                  # Append ref n
...
```
You can read more about them here https://www.nushell.sh/book/custom_completions.html#example-custom-completion

**Reply by wmstack** [reply]
I see, so one has to do `git fetch` and then press tab, and then nushell runs the "nu-complete git remotes" command to extract the possible completions.

**Reply by fdncred** [reply]
Yes, exactly. one last thing, the names like "nu-complete git remotes" are just names. I'm pretty sure there's no secret sauce with the term "nu-complete". It could probably be named "do-some-cool-thing-with" and it should work fine as long as it's referenced by name. It's just a custom command name.
